adc/open.py

```diff
@@ -0,0 +1,40 @@
+from typing import Union, Iterable, List
+import confluent_kafka
+
+from adc import producer, consumer, kafka
+
+
+def open(url: str, mode: str = 'r') -> Union[producer.Producer, Iterable[confluent_kafka.Message]]:
```

Comment:
Yes, that's basically true. I'm not sure how to make the `open` interface both clear _and_ powerful. `**kwargs` obscures everything in a way I really don't like... maybe it can accept a `consumer_config` / `producer_config`? It's weird, though, if you open in `r` mode but provide a `producer_config`.

Reply:
There should be an analog to the CAP theorem on this, where you can only pick one out of highly configurable and ease-of-use. I agree with your sentiment there, and wonder if we should really be exposing configuration-level options like this in the first place?

For my use case, I only ever care about needing to make configuration-level changes for (1) offset-based options and (2) auth-based options. Before having the auth classes, we needed to expose these config options to be able to provide that but now we don't. For (1), that was something I needed to do in the stock client but here I don't see a need to make changes. I would be good just exposing an `auth` option. If we really need to expose options, maybe we can expose higher level functionality that make the needed changes as that use case arises?

Reply:
Please take a look at the new version. It's a bit of a mess, but it exposes a bunch of things. It's kind of nasty how we have to just warn when reader-side options are passed in during write mode, but it is what it is.

I think any sophisticated application should probably avoid `open` and construct Consumers and Producers explicitly.